### PR TITLE
Add harness tests for array list proofs

### PR DESCRIPTION
**Pull request by feliperodri**
Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

*Description of changes:*

This PR introduces a collection of harness tests to proof memory safety of the `aws_array_list` implementation. Each harness contains: 

* an arbitrary `aws_array_list`, which is initialized either dynamically or statically;
* a single operation (e.g., `aws_array_list_back` or `aws_array_list_pop_front`);
* some functional guarantees described as assertions;

All proofs are based on the set of assumptions described at the header of each harness function.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_array_list_back/aws_array_list_back_harness.c [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/array_list.h>
+#include <proof_helpers/make_common_data_structures.h>
+
+#define MAX_ITEM_SIZE 2
+#define MAX_INITIAL_ITEM_ALLOCATION (UINT64_MAX / MAX_ITEM_SIZE) + 1
```
Comment (markrtuttle) [inline .cbmc-batch/jobs/aws_array_list_back/aws_array_list_back_harness.c]: Why is this a natural choice?  Seems like a lot of elements to allocate in a CBMC test.

File: .cbmc-batch/jobs/aws_array_list_back/aws_array_list_back_harness.c [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    size_t item_size = list->item_size;
+    void *data = list->data;
+
+    if (!aws_array_list_back(list, val)) {
+        /* some guarantees */
+        assert(list->data);
+        assert(list->length);
+    }
+
+    /* some guarantees */
```
Comment (markrtuttle) [inline .cbmc-batch/jobs/aws_array_list_back/aws_array_list_back_harness.c]: Do you mean a guarantee that the list is unchanged?

File: .cbmc-batch/jobs/aws_array_list_front/aws_array_list_front_harness.c [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ *     - non-deterministic item_size <= MAX_ITEM_SIZE
+ *     - non-deterministic length <= initial_item_allocation
+ *     - aws_array_list_front == AWS_OP_SUCCESS
+ *                                 -> list->data != NULL
+ *                                 -> list->length != 0
+ *     - All parameters remain the same after this operation
+ */
+void aws_array_list_front_harness() {
+    struct aws_array_list *list;
+    ASSUME_BOUNDED_ARRAY_LIST(list, MAX_INITIAL_ITEM_ALLOCATION, MAX_ITEM_SIZE);
```
Comment (markrtuttle) [inline .cbmc-batch/jobs/aws_array_list_front/aws_array_list_front_harness.c]: This (here and elsewhere) appears to be the only assumption you are making, does it encode all of the assumptions listed a few lines above?

File: .cbmc-batch/jobs/aws_array_list_get_at/aws_array_list_get_at_harness.c [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ *                                 -> list->data != NULL
+ *                                 -> list->length > index
+ *     - all parameters remain the same after this operation
+ */
+void aws_array_list_get_at_harness() {
+    struct aws_array_list *list;
+    ASSUME_BOUNDED_ARRAY_LIST(list, MAX_INITIAL_ITEM_ALLOCATION, MAX_ITEM_SIZE);
+
+    void *val = malloc(list->item_size);
+    size_t index = nondet_size_t();
```
Comment (markrtuttle) [inline .cbmc-batch/jobs/aws_array_list_get_at/aws_array_list_get_at_harness.c]: Why invoke a function to assign a nondet value when CBMC itself will assign a nondet value to an uninitialized local?

File: .cbmc-batch/jobs/aws_array_list_init_dynamic/aws_array_list_init_dynamic_harness.c [hunk 5]
```diff
@@ -1,21 +1,58 @@
...
+ *                                    || list->data &&
+ *                                       list->current_size == initial_item_allocation * item_size
+ */
 void aws_array_list_init_dynamic_harness() {
     struct aws_array_list *list;
     ASSUME_VALID_MEMORY(list);
     struct aws_allocator *allocator;
-    ASSUME_DEFAULT_ALLOCATOR(allocator);
-    size_t item_count = nondet_size_t();
+    ASSUME_CAN_FAIL_ALLOCATOR(allocator);
```
Comment (markrtuttle) [inline .cbmc-batch/jobs/aws_array_list_init_dynamic/aws_array_list_init_dynamic_harness.c]: The modeling style changes here.  Before the assumption that an allocator can fail was a parameter to a function defining the list.  Here the list and the allocator are treated separately.

File: .cbmc-batch/jobs/aws_array_list_init_dynamic/aws_array_list_init_dynamic_harness.c [hunk 6]
```diff
@@ -1,21 +1,58 @@
...
+    size_t initial_item_allocation = nondet_size_t();
+    __CPROVER_assume(initial_item_allocation <= MAX_INITIAL_ITEM_ALLOCATION);
     size_t item_size = nondet_size_t();
```
Comment (markrtuttle) [inline .cbmc-batch/jobs/aws_array_list_init_dynamic/aws_array_list_init_dynamic_harness.c]: Here and elsewhere, and mentioned above, this invocation of nondet_size_t is not necessary, cbmc will do this on its own.

File: .cbmc-batch/jobs/aws_array_list_back/aws_array_list_back_harness.c (hunk 2 above)
Comment (feliperodri) [reply .cbmc-batch/jobs/aws_array_list_back/aws_array_list_back_harness.c]: Yes, I can simply replace this by /* assertions */

File: .cbmc-batch/jobs/aws_array_list_front/aws_array_list_front_harness.c (hunk 3 above)
Comment (feliperodri) [reply .cbmc-batch/jobs/aws_array_list_front/aws_array_list_front_harness.c]: Yes, should I add an extra comment to clarify this here?

File: .cbmc-batch/jobs/aws_array_list_get_at/aws_array_list_get_at_harness.c (hunk 4 above)
Comment (feliperodri) [reply .cbmc-batch/jobs/aws_array_list_get_at/aws_array_list_get_at_harness.c]: For the sake of clarity for those who are not familiar with CBMC. Should I remove it?

File: .cbmc-batch/jobs/aws_array_list_init_dynamic/aws_array_list_init_dynamic_harness.c (hunk 5 above)
Comment (feliperodri) [reply .cbmc-batch/jobs/aws_array_list_init_dynamic/aws_array_list_init_dynamic_harness.c]: This particular harness is design to proof memory safety for the function `aws_array_list_init_dynamic`, which dynamically initiates an `aws_arra_list` structure. For other harnesses, we need to non-deterministically create an `aws_array_list`, which could be initialized either statically or dynamically, thus, the use of `ASSUME_BOUNDED_ARRAY_LIST` in other cases. That's why here we treat allocator separately.

**Pull request merged**